src/Microsoft.TestPlatform.Common/Filtering/FastFilter.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.Common.Filtering
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Text.RegularExpressions;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+    internal sealed class FastFilter
+    {
+        internal string FilterPropertyName { get; }
+
+        internal ImmutableHashSet<string> FilterPropertyValues { get; }
+
+        internal bool IsFilteredOutWhenMatched { get; }
+
+        internal Regex PropertyValueRegex { get; set; }
+
+        internal FastFilter(string filterPropertyName, ImmutableHashSet<string> filterPropertyValues, Operation filterOperation, Operator filterOperator)
+        {
+            ValidateArg.NotNullOrEmpty(filterPropertyName, nameof(filterPropertyName));
+            ValidateArg.NotNull(filterPropertyValues, nameof(filterPropertyValues));
+            
+            this.FilterPropertyName = filterPropertyName;
+            this.FilterPropertyValues = filterPropertyValues;
+
+            if (filterOperation == Operation.Equal && (filterOperator == Operator.Or || filterOperator == Operator.None))
+            {
+                IsFilteredOutWhenMatched = false;
+            }
+            else if (filterOperation == Operation.NotEqual && (filterOperator == Operator.And || filterOperator == Operator.None))
+            {
+                IsFilteredOutWhenMatched = true;
+            }
+            else
+            {
+                throw new ArgumentException();
+            }
+        }
+
+        internal string[] ValidForProperties(IEnumerable<string> properties)
+        {
+            return properties.Contains(this.FilterPropertyName, StringComparer.OrdinalIgnoreCase)
+                ? null
+                : new[] { this.FilterPropertyName };
+        }
+
+        internal bool Evaluate(Func<string, Object> propertyValueProvider)
+        {
+            ValidateArg.NotNull(propertyValueProvider, "propertyValueProvider");
+
+            if (!TryGetSinglePropertyValue(this.FilterPropertyName, propertyValueProvider, out var value))
+            {
+                return false;
+            }
+
+            if (PropertyValueRegex != null)
```

Comment:
nit: Since `Evaluate` would be called for every test case, this if and the `IsFilteredOutWhenMatched` condition could be avoided if the implementation changes to have an inheritance with a Factory to return back the right FastFilter instance. Might just make it faster. Thoughts?

Reply:
well, I don't think the logic and overhead here warrants an abstraction like that yet.